enhancements/machine-config/admin-defined-reboot-policy.md

```diff
@@ -0,0 +1,264 @@
+
+## Motivation
+
+Historically, workload disruption (drains and reboots) can be very costly for OCP users, especially in bare-metal environments where the stopping and restarting of a node can take up to an hour. For small incremental changes to the host OS, many of them can take effect with a service reload or no action at all, and thus the ability to perform a non-disruptive update can be very beneficial to many users without any downside.                  
+
+The MCO since OCP 4.7 has added the ability to perform disruptionless updates on select changes, see: [MCO doc on disruptionless updates](https://github.com/openshift/machine-config-operator/blob/master/docs/MachineConfigDaemon.md#rebootless-updates). This list however is by no means exhaustive, and it is not feasible for the MCO to hard-code all use cases for all users into the MCO. In addition, different users may have different expectations on what will cause a reboot or not. Thus there is a need for a user-defined, MCO operated list of "reboot policies" on what to do after an OS update has been applied.
+
+### User Stories
+
+* As an Openshift cluster administrator, I have an environment with thousands of ICSP/IDMS mirror entries. These entries will be added and deleted constantly on a weekly basis. I understand the potential dangers of removing mirrors (can cause an image to no longer be accessible if no other mirror hosts it), but I accept that risk and would like my changes to apply without disrupting my workloads and only by reloading the crio service.
+
+* As an application developer, I have certificates for my applications I need on-disk, which can be rotated safely and hot-loaded into the applications without any additional action. I do not wish to disrupt other applications when these update via the MCO.
+
+### Goals
+
+* Create a new RebootPolicy CRD to allow users to define how OS updates are actioned on after application
+* Have users be able to define no-action and service reloads to specific MachineConfig changes
+* Have users be able to easily see existing cluster non-disruptive update cases
+
+### Non-Goals
+
+* Have RebootPolicy apply to non-MCO driven changes (e.g. SRIOV can still reboot nodes)
+* Remove existing non-disruptive update paths (the user will be able to override cluster defaults)
+* Design for image-based updates (live apply and bootc, will be considered in the future)
+* Have the MCO validate whether a change can be successfully applied with the given RebootPolicy (i.e. it is up to the responsibility of the user to ensure the correctness of their defined actions)
+
+## Proposal
+
+Create a new rebootpolicy.machineconfiguration.openshift.io CRD, allowing specify user-provided rebootpolicies, and view current version cluster defaults, as follows:
+
+```console
+apiVersion: machineconfiguration.openshift.io/v1
+kind: RebootPolicy
```

Comment:
So this enhancement doesn't yet mention the [bootc configmap](https://github.com/containers/bootc/issues/22) approach (we aren't using it yet I know, doesn't really exist)

But anyways in there I was straw-manning that we have an annotation in the configmap which specifies the "update policy" for it (apply live vs reboot) - there's some ugly bits around whether we need "apply live and execute this arbitrary command" or "apply live and reload these systemd units" that need to be defined.

But anyways to me, associating the update policy with the files makes a lot more sense than having another global list.

Now to *implement* this today we'd need to add a "reboot policy" section into the rendered config.  But I think it would be a lot nicer in the future to get rid of the idea of the big "rendered config" and have the MCD fetch individual MachineConfig objects (like how kubelet fetches individual configmaps, there's no "rendered configmap!").

Doing that gets into versioning/race condition questions around "what happens when an input MC changes while we're doing a rollout" that I think are solvable, but do need some thought.

Now one thing @yuqi-zhang brought up when we were chatting about this is the overall question of "how would this work for files in the base image".

My offhand thoughts on that are that we may want to go in a direction of having "static configmaps" that we can inject into a container image in some way, similar to the arguments around https://github.com/containers/bootc/issues/128

(These things connect because configmaps may specify quadlet units)

Then `bootc` would be the thing that executes on applying these note!

Anyways...I'm not totally sure we really need to immediately handle the case of handling policies for files embedded in the base image.

It feels like we'd get most of the way if we split out the things that need to be updated live outside of the base image.

Reply:
>  I did consider this approach and originally rejected it because most of the "no-reboot updates" that are asked about are certificates, or registry configuration etc. which comes all in the cluster-supplied 00-worker and 00-master MCs, which you can't edit.

But we can easily split out those files into a distinct e.g. `01-worker-configs` MC today and aim for stuff that changes in -configs to have distinct "apply" policies.